File: Python/fibonacci.py
from time import time

#Recursion
def recursion(n):
	if n <= 1:
		return n
	return recursion(n - 1) + recursion(n - 2)

#Dynamic programming	
def dynamic(n):
	l = [0, 1]
	if n > 1:
		for i in range(2, n + 1):
			l.append(l[i - 1] + l[i - 2])
	return l[n]

#Space optimization method	
def space_optimized(n):
	a, b = 0, 1
	if n > 1:
		for i in range(2, n + 1):
			c = a + b
			a, b = b, c
	return b

def main():
	while(1):
		print("""
[1] Recursion
[2] Dynamic programming
[3] Space optimization method
[4] Exit""")
		choice = int(input("Enter choice : "))
	
		if choice == 1:
			print("Fibonacci using Recursion")
			n = int(input("Enter n : "))
			start = time()
			fib = recursion(n)
			end = time()
			print(f"{n} fibonacci series is {fib}")
			print("Program run time ",end - start)
		
		elif choice == 2:
			print("Fibonacci using Dynamic programming ")
			n = int(input("Enter n : "))
			start = time()
			fib = dynamic(n)
			end = time()
			print(f"{n} fibonacci series is {fib}")
			print("Program run time ",end - start)
		
		elif choice == 3:
			print("Fibonacci using Space optimization method")
			n = int(input("Enter n : "))
			start = time()
			fib = space_optimized(n)
			end = time()
			print(f"{n} fibonacci series is {fib}")
			print("Program run time ",end - start)
		
		elif choice == 4:
			break
		
		else:
			print("Invalid choice")
			
if __name__ == "__main__":
	main()

File: C/fibonacci.c
#include<stdio.h>
#include<time.h>
//#include<stdlib.h>


//Recursive function
int recursion(int n){
  if(n<=1){
    return n;
  }
  return recursion(n-1) + recursion(n-2);
}


//Dynamic programming
int dynamic(int n){
  int arr[n + 1];
  int i;

  arr[0] = 0;
  arr[1] = 1;

  for(i = 2; i <= n; i++){
    arr[i] = arr[i -1] + arr[i -2];
  }
  return arr[n];
}


//Space optimized method
int space_optimized(int n){
	int a = 0, b = 1, i, c;
	if(n == 0){
		return a;
	}
	for(i = 2;i <= n;i++){
		c = a + b;
		a = b;
		b = c;
	}
	return b;
}


int main(){

clock_t start, end;
double time;
int choice, n, fib;
printf("Fibonacci series\n\n");

while(1){
  printf("[1] Recursion\n[2] Dynamic Programming\n[3] Space Optimized Method\n[4] Exit\n\nEnter choice : ");
  scanf("%d", &choice);
  
  if(choice == 1){
    printf("Fibonacci using recursion\n\nEnter n : ");
    scanf("%d", &n);
    start = clock();
    fib = recursion(n);
    end = clock() - start;
    time = ((double)end) / CLOCKS_PER_SEC;
    printf("%d Fibonacci series is %d\n", n, fib);
    printf("\nTime taken %f\n\n ", time);

  }
  else if(choice == 2){
    printf("Fibonacci using dynamic programming\n\nEnter n : ");
    scanf("%d", &n);
    start = clock();
    fib = dynamic(n);
    end = clock() - start;
    time = ((double)end) / CLOCKS_PER_SEC;
    printf("%d Fibonacci series is %d\n", n, fib);
    printf("\nTime taken %f\n\n", time);

  }
  else if(choice == 3){
    printf("Fibonacci series using Space optimized program\n\nEnter n : ");
    scanf("%d", &n);
    start = clock();
    fib = space_optimized(n);
    end = clock() - start;
    time = ((double)end) / CLOCKS_PER_SEC;
    printf("%d Fibonacci series is %d\n", n, fib);
    printf("\nTime taken %f\n\n", time);
  }
  else if(choice == 4){
    //exit(0);
    break;
  }
  else{
    printf("Invalid choice\n\n");
  }
}
return 0;
}
